src/dlls/mscoree/mscoree.cpp

```diff
@@ -1263,17 +1237,27 @@ HRESULT SetInternalSystemDirectory()
         DWORD len;
 
         // use local buffer for thread safety
-        WCHAR wzSystemDirectory[COUNTOF(g_pSystemDirectory)];
+        WCHAR * wzSystemDirectory = new (nothrow) WCHAR[MAX_LONGPATH+1];
+        if (wzSystemDirectory == NULL)
+        {
+            WCHAR wzSystemDirectoryStack[MAX_LONGPATH+1];
+            wzSystemDirectory = wzSystemDirectoryStack;
+        }
         
-        hr = GetCORSystemDirectoryInternal(wzSystemDirectory, COUNTOF(wzSystemDirectory), &len);
+        hr = GetCORSystemDirectoryInternal(wzSystemDirectory, MAX_LONGPATH+1, &len);
 
         if(FAILED(hr)) {
-            wzSystemDirectory[0] = W('\0');
+            *wzSystemDirectory = W('\0');
             len = 1;
         }
-
+        
         // publish results idempotently with correct memory ordering
         memcpy(g_pSystemDirectory, wzSystemDirectory, len * sizeof(WCHAR));
```

Comment:
Your original code would allocate MAX_PATHLONG buffer for the duration of the program, that will be mostly unused in the typical case (the typical path will be much smaller than MAX_PATHLONG). This will avoid wasting memory - the buffer allocated for the duration of the program will be of just the right size.